collections/src/commonMain/kotlin/AtomicMap.kt

```diff
@@ -0,0 +1,94 @@
+package com.juul.tuulbox.collections
+
+import kotlinx.collections.immutable.ImmutableCollection
+import kotlinx.collections.immutable.ImmutableMap
+import kotlinx.collections.immutable.ImmutableSet
+import kotlinx.collections.immutable.PersistentMap
+import kotlinx.collections.immutable.mutate
+import kotlinx.collections.immutable.persistentHashMapOf
+import kotlinx.collections.immutable.persistentMapOf
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.getAndUpdate
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.flow.updateAndGet
+
+/** Returns an [AtomicMap] that guarantees preservation of iteration order, but may be slower. */
+public fun <K, V> atomicMapOf(
+    vararg pairs: Pair<K, V>,
+): AtomicMap<K, V> = AtomicMap(persistentMapOf(*pairs))
+
+/** Returns an [AtomicMap] that does not guarantee preservation of iteration order, but may be faster. */
+public fun <K, V> atomicHashMapOf(
+    vararg pairs: Pair<K, V>,
+): AtomicMap<K, V> = AtomicMap(persistentHashMapOf(*pairs))
+
+/**
+ * A [Map] that allows for thread safe, atomic mutation. Returned collections such as [entries] and
+ * [iterator] reference a [snapshot] of when they were accessed, and are not mutated when the map is.
+ *
+ * Although mutable, this class intentionally does not implement [MutableMap]. Mutation must use
+ * designated mutator functions ([mutate], [snapshotAndMutate], [mutateAndSnapshot]).
+ */
+public class AtomicMap<K, V> private constructor(
+    @PublishedApi
+    internal val state: MutableStateFlow<PersistentMap<K, V>>,
+) : Map<K, V> {
+
+    internal constructor(initial: PersistentMap<K, V>) : this(MutableStateFlow(initial))
+
+    /** Returns this map as a [StateFlow]. Each call to [put], [mutate], etc will cause a new emit on this flow. */
+    public val flow: StateFlow<ImmutableMap<K, V>> = state.asStateFlow()
+
+    /**
+     * Returns the current value of this map as an [immutable][ImmutableMap] snapshot.
+     *
+     * This operation is non-copying and efficient.
+     */
+    public val snapshot: ImmutableMap<K, V>
+        get() = state.value
+
+    override val size: Int
+        get() = snapshot.size
+
+    override val entries: ImmutableSet<Map.Entry<K, V>>
+        get() = snapshot.entries
+
+    override val keys: ImmutableSet<K>
+        get() = snapshot.keys
+
+    override val values: ImmutableCollection<V>
+        get() = snapshot.values
+
+    override fun containsKey(key: K): Boolean = snapshot.containsKey(key)
+
+    override fun containsValue(value: V): Boolean = snapshot.containsValue(value)
+
+    override fun get(key: K): V? = snapshot[key]
+
+    override fun isEmpty(): Boolean = snapshot.isEmpty()
+
+    override fun equals(other: Any?): Boolean = snapshot == other
```

Comment:
Do you happen to know if the immutable collections do content equality checks, or is this a reference equality check?

Reply:
They compare contents. It's actually a Java requirement that Kotlin inherited.

https://javadoc.scijava.org/Java/java/util/Map.html#equals-java.lang.Object-

> Compares the specified object with this map for equality. Returns true if the given object is also a map and the two maps represent the same mappings. More formally, two maps m1 and m2 represent the same mappings if m1.entrySet().equals(m2.entrySet()). This ensures that the equals method works properly across different implementations of the Map interface.

Reply:
Awesome, thanks for clarifying!